tsdb/agent/db_test.go

```diff
@@ -934,6 +936,223 @@ func TestDBOutOfOrderTimeWindow(t *testing.T) {
 	}
 }
 
+type walSample struct {
+	t    int64
+	f    float64
+	h    *histogram.Histogram
+	lbls labels.Labels
+	ref  storage.SeriesRef
+}
+
+func TestDBCreatedTimestampSamplesIngestion(t *testing.T) {
+	t.Parallel()
+
+	type appendableSample struct {
+		t            int64
+		ct           int64
+		v            float64
+		lbls         labels.Labels
+		h            *histogram.Histogram
+		expectsError bool
+	}
+
+	testHistogram := tsdbutil.GenerateTestHistograms(1)[0]
+	zeroHistogram := &histogram.Histogram{}
+
+	invalidHist := &histogram.Histogram{
+		CustomValues: make([]float64, 0, 1),
+	}
+
+	lbls := labelsForTest(t.Name(), 1)
+	defLbls := labels.New(lbls[0]...)
+
+	testCases := []struct {
+		name                string
+		inputSamples        []appendableSample
+		expectedSamples     []*walSample
+		appendFunc          func(app storage.Appender, lset labels.Labels, t, ct int64, value interface{}) (storage.SeriesRef, error)
+		appendErrorFunc     func(app storage.Appender, ref storage.SeriesRef, lset labels.Labels, t, ct int64, value interface{}) error
+		value               interface{}
+		expectedSampleCount int
+		expectedType        string
+		expectedErrorMsg    string
+		expectedSeriesCount int
+	}{
+		{
+			name: "CT+float && CT+histogram samples",
+			inputSamples: []appendableSample{
+				{
+					t:    100,
+					ct:   30,
+					v:    0,
+					lbls: defLbls,
+				},
+				{
+					t:    200,
+					v:    20,
+					lbls: defLbls,
+				},
+				{
+					t:    300,
+					ct:   230,
+					h:    zeroHistogram,
+					lbls: defLbls,
+				},
+				{
+					t:    400,
+					h:    testHistogram,
+					lbls: defLbls,
+				},
+			},
+			expectedSamples: []*walSample{
+				{t: 30, f: 0, lbls: defLbls},
+				{t: 100, f: 20, lbls: defLbls},
+				{t: 230, h: zeroHistogram, lbls: defLbls},
+				{t: 300, h: testHistogram, lbls: defLbls},
+			},
+			expectedSeriesCount: 1,
+		},
+		{
+			name: "CT+float && CT+histogram samples with error",
+			inputSamples: []appendableSample{
+				{
+					// invalid CT
+					t:            100,
+					ct:           100,
+					v:            0,
+					lbls:         defLbls,
+					expectsError: true,
+				},
+				{
+					// invalid Histogram, will validate
+					t:            200,
+					h:            invalidHist,
+					lbls:         defLbls,
+					expectsError: true,
+				},
+				{
+					// invalid CT histogram
+					t:            300,
+					ct:           300,
+					h:            zeroHistogram,
+					lbls:         defLbls,
+					expectsError: true,
+				},
+			},
+			expectedSeriesCount: 0,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			reg := prometheus.NewRegistry()
+			opts := DefaultOptions()
+			opts.OutOfOrderTimeWindow = 0
+			s := createTestAgentDB(t, reg, opts)
+			app := s.Appender(context.TODO())
+
+			for _, sample := range tc.inputSamples {
+				// We supposed to write a Histogram to the WAL
+				if sample.h != nil {
+					if sample.ct != 0 {
```

Comment:
We reviewed together in CT work group and we still would love consistency with the other [testl](https://github.com/prometheus/prometheus/blob/2cabd1b707b0569f65fbc2bafa899cabf47cfe09/tsdb/head_test.go#L6283-L6502) @ArthurSens linked. It feels we should always append both CT and sample?

Reply:
So, now after I read the code from scrape.go I know what you guys meant by that. I am catching up on the whole CT development here, so sorry for the fumbles. 

By following the behavior from `scrape.go` I hit a failure in my test which is legit. Going to ask the question here as well (as I did in Slack). Going to follow up wherever I get answer first:

> Guys, one question I have about CT + PRW in the agent.
> 
> In the agent the Commit() for the appender is quite primitive, it just encodes all samples and series and put on WAL. Now that we are implementing support on the agent, should we replicate the logic from the head appender's Commit() where we check for valid CT or should we try somehow to validate it at Append() time?
> 
> Or, should we leave to the receiver of the PRW to decide what to do with the timestamp, from what I understand it will flag as OOO and drop it. More context here: https://github.com/prometheus/prometheus/pull/15124#discussion_r1812822048

Reply:
Okay, I tackled this last problem with the help of bartek in slack. Here is the message:

> Yea, so in the soon future it will be receiver because it will part of remote write request, as a separate field (it's already allowed by proto, we just need to put that in the record)
> For now, it depends on ooo flow, but prw does not allow ooo in the batch so one way or another this can't just rely on receiver OR we need special batching flow on sender which is pretty wasteful for this case (:

I am handling the duplicate CT samples and ignoring them (`ErrOutOfOrderCT`) by keeping track of lastTs for the series before Commit() is called.